### PR TITLE
Chain register

### PR DESCRIPTION
**Pull request by marcelovani**
**ClassRegister**
This is a proof of concept.
I would like to be able to call chain commands without having to specify --file or other arguments, because it takes too much time to write them. I would like to just type

> drupal chain:someting

Also, since site:new is not currently available, I would like to have it as a chain command

> drupal chain:site:new
This would be the content of the chain-site-new.yml
commands:
# Build from template
```
  - command: exec
    arguments:
      bin: 'composer create-project %{{project|dennisdigital/drupal-project:8.x-dev}}:%{{version|8.x-dev}} %{{directory|/vagrant/repos/new-site}} --stability dev --no-interaction'
```
Say I want to put this on jenkins, I would have to do

> drupal chain --file=/pathtofile/chain-site-new.yml --parameter="profile:profile_name" --parameter="directory:/somedir" --parameter="version:8.1"

With this prototype, I can skip the --file

Another possibility is to have a chain calling another chain.
chain-site-test.yml
```
commands:
# Example of a chain calling another chain with placeholders
  - command: exec
    arguments:
      bin: 'drupal chain:site:test-child --placeholder="project:foo" --placeholder="directory:some-dir" --placeholder="version:10"'
```

chain-site-test-child.yml
commands:
```
# This chain shows the variables passed either via placeholders or interactive
  - command: exec
    arguments:
      bin: 'echo "This is a test %{{project|dennisdigital/drupal-project:8.x-dev}} %{{directory|/vagrant/repos/new-site}} %{{version|8.x-dev}}"'
```

I can run drupal chain:site:test and it will cascade the parameters

> the output will be
> This is a test foo some-dir 10

How does it work:
Copy your chain commands to any folder
Place the chain.yml in the composer folder
# Custom chains
```
chain:
  name:
    'site:new':
      file: '~/.config/composer/dev_scripts/vendor/dennisdigital/drupal_console_commands/config/chain/chain-site-new.yml'
'~/.config/composer/dev_scripts/vendor/dennisdigital/drupal_console_commands/config/chain/chain-site-install.yml'
  disabled:
    'site:test':
      file: '~/.composer/dev_scripts/vendor/dennisdigital/drupal_console_commands/config/chain/chain-site-test.yml'
```

### CODE REVIEW

**Comment by marcelovani** [comment]
I merged the latest changes from master to my branch. The code isn't working now, because Config.php was deleted and I don't know how to point to chain.yml

File: chain.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Chains listed here will be auto loaded by the ChainRegister.
```
Comment (jmolivas) [inline chain.yml]: This file is no longer required on this PR

File: src/Application.php [hunk 2]
```diff
@@ -107,6 +108,24 @@ private function registerCommands()
...
+          )
+        );
+
+        foreach ($chainCommands as $name => $chainCommand) {
+            $file = $chainCommand['file'];
+            $command = new ChainRegister($name, $file);
+            $this->add($command);
+        }
+
+        $autoWireForcedCommands = $this->getConfig()->get(
```
Comment (jmolivas) [inline src/Application.php]: can you remove from 123 to 127 those lines got relocated to drupal-console-core

File: src/Application.php [hunk 3]
```diff
@@ -107,6 +108,24 @@ private function registerCommands()
             $logger->writeln($this->trans('application.site.errors.settings'));
         }
 
+        $chainCommands = $this->getConfig()->get(
```
Comment (jmolivas) [inline src/Application.php]: can you remove from 111 to 115 the loading commands will not be reading a config file.
I will write the implementation

**Comment by marcelovani** [comment]
Moved PR to the sponsoring company github account https://github.com/hechoendrupal/DrupalConsole/pull/2961/files

**Pull request closed without merging**